core/src/spotify_uri.rs

```diff
@@ -0,0 +1,321 @@
+use crate::spotify_id::SpotifyItemType;
+use crate::{Error, SpotifyId};
+use std::fmt;
+use thiserror::Error;
+
+use librespot_protocol as protocol;
+
+#[derive(Debug, Error, Clone, Copy, PartialEq, Eq)]
+pub enum SpotifyUriError {
+    #[error("not a valid Spotify URI")]
+    InvalidFormat,
+    #[error("URI does not belong to Spotify")]
+    InvalidRoot,
+}
+
+impl From<SpotifyUriError> for Error {
+    fn from(err: SpotifyUriError) -> Self {
+        Error::invalid_argument(err)
+    }
+}
+
+pub type SpotifyUriResult = Result<SpotifyUri, Error>;
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum SpotifyUri {
+    Album {
+        id: SpotifyId,
+    },
+    Artist {
+        id: SpotifyId,
+    },
+    Episode {
+        id: SpotifyId,
+    },
+    Playlist {
+        user: Option<String>,
+        id: SpotifyId,
+    },
+    Show {
+        id: SpotifyId,
+    },
+    Track {
+        id: SpotifyId,
+    },
+    Local {
+        artist: String,
+        album_title: String,
+        track_title: String,
+        duration: std::time::Duration,
+    },
+    Unknown {
+        id: SpotifyId,
```

Comment:
Shouldn't this rather be a String which represents the whole uri? Otherwise we would lose all info about what type was unknown.